File: leetcode/stack/402. Remove K Digits/js/main.js
// Mono Stack

// Time O(N)
// Space O(N)
const removeKdigits = (num, k) => {
  let stack = [];
  let n = num.length;

  for (let i = 0; i < n; i++) {
    while (k > 0 && stack.length && last(stack) > num[i]) {
      stack.pop();
      k--;
    }

    stack.push(num[i]);
  }

  while (k > 0) {
    stack.pop();
    k--;
  }

  while (stack[0] == '0' && stack.length > 0) {
    stack.shift();
  }

  return stack.length == 0 ? '0' : stack.join('');

  function last(x) {
    return x[x.length - 1];
  }
};

File: leetcode/stack/402. Remove K Digits/py/main.py
import collections 

# Time O(N)
# Space O(N)
class Solution:
    def removeKdigits(self, num, k):
        if len(num) == k: 
          return '0'

        st, N = collections.deque(), len(num)

        for i in range(N):
          x = int(num[i])
          while st and st[-1] > x and k > 0:
            st.pop()
            k -= 1
          
          st.append(x)
        
        while k > 0:
            st.pop()
            k -= 1
        
        while len(st) > 1 and st[0] == 0:
          st.popleft()

        return ''.join([str(val) for val in st]) 
        